sound/soc/sof/intel/hda-dai-ops.c

```diff
@@ -155,11 +155,123 @@ static void hda_dma_free(struct snd_sof_dev *sdev, void *dma_data)
 	snd_hdac_ext_stream_reset(hext_stream);
 }
 
+static int hda_ipc4_pre_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+				struct snd_pcm_substream *substream, int cmd)
+{
+	struct snd_sof_widget *pipe_widget;
+	struct sof_ipc4_pipeline *pipeline;
+	struct snd_sof_widget *swidget;
+	struct snd_soc_dapm_widget *w;
+	int ret;
+
+	w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
+	swidget = w->dobj.private;
+	pipe_widget = swidget->pipe_widget;
```

Comment:
This will be conflict with the `ASoC: SOF: Introduce struct snd_sof_pipeline` commit